lib/Dialect/FIRRTL/Transforms/LowerBundleVectorTypes.cpp

```diff
@@ -0,0 +1,967 @@
+  if (processSAPath(op))
+    return;
+
+  SmallVector<FlatBundleFieldEntry> srcFields = peelType(op.src().getType());
+  SmallVector<FlatBundleFieldEntry> destFields = peelType(op.dest().getType());
+
+  // Ground Type
+  if (destFields.empty()) {
+    // check for truncation
+    Value src = op.src();
+    Value dest = op.dest();
+    auto srcType = src.getType().cast<FIRRTLType>();
+    auto destType = dest.getType().cast<FIRRTLType>();
+    auto srcWidth = srcType.getBitWidthOrSentinel();
+    auto destWidth = destType.getBitWidthOrSentinel();
+
+    if (destType == srcType) {
+      builder->create<ConnectOp>(dest, src);
+      opsToRemove.push_back(op);
+    } else if (destType.isa<IntType>() && srcType.isa<IntType>() &&
+               destWidth >= 0 && destWidth < srcWidth) {
+      // firrtl.tail always returns uint even for sint operands.
+      IntType tmpType = destType.cast<IntType>();
+      if (tmpType.isSigned())
+        tmpType = UIntType::get(destType.getContext(), destWidth);
+      src = builder->create<TailPrimOp>(tmpType, src, srcWidth - destWidth);
+      // Insert the cast back to signed if needed.
+      if (tmpType != destType)
+        src = builder->create<AsSIntPrimOp>(destType, src);
+      builder->create<ConnectOp>(dest, src);
+      opsToRemove.push_back(op);
+    }
+    return;
+  }
+
+  // Aggregates
+  if (FVectorType fvector = op.src().getType().dyn_cast<FVectorType>()) {
+    for (int index = 0, e = std::min(srcFields.size(), destFields.size());
+         index != e; ++index) {
+      Value src = builder->create<SubindexOp>(op.src(), index);
+      Value dest = builder->create<SubindexOp>(op.dest(), index);
+      if (src.getType() == dest.getType())
+        builder->create<ConnectOp>(dest, src);
+      else
+        builder->create<PartialConnectOp>(dest, src);
+    }
+  } else if (BundleType srcBundle = op.src().getType().dyn_cast<BundleType>()) {
+    // Pairwise connect on matching field names
+    BundleType destBundle = op.dest().getType().cast<BundleType>();
+    for (int srcIndex = 0, srcEnd = srcBundle.getNumElements();
+         srcIndex < srcEnd; ++srcIndex) {
+      auto srcName = srcBundle.getElement(srcIndex).name;
+      for (int destIndex = 0, destEnd = destBundle.getNumElements();
+           destIndex < destEnd; ++destIndex) {
+        auto destName = destBundle.getElement(destIndex).name;
+        if (srcName == destName) {
+          Value src = builder->create<SubfieldOp>(op.src(), srcName);
+          Value dest = builder->create<SubfieldOp>(op.dest(), destName);
+          if (destFields[destIndex].isOutput)
+            std::swap(src, dest);
+          if (src.getType().isa<AnalogType>())
+            builder->create<AttachOp>(ArrayRef<Value>{dest, src});
+          else if (src.getType() == dest.getType())
+            builder->create<ConnectOp>(dest, src);
+          else
+            builder->create<PartialConnectOp>(dest, src);
+          break;
+        }
+      }
+    }
```

Comment:
It would be nice if there was some utility that avoided the `O(n^2)` connection here. Maybe drop a `TODO`.

Reply:
I thought of this, the obvious choice is to sort and intersect.  I didn't because this has a large fixed overhead and I believe that types will generally be small enough that this should be faster than something more complex.